File: site/src/subdomains/users/valueobjects.ts
import { VO } from './../../common/vo/vo';

// name
// password
// login

const nameRegex = new RegExp("^[a-zA-Z0-9-_]*$")

const toShortNameError = new Error("name has to be at least 3 characters long")
const toLongNameError = new Error("name cannot have more than 64 characters")
const invalidCharachtersNameError = new Error("name can contain only letters, numbers, - and _")

export class Name implements VO {
  public Value: string

  constructor(value: string) { this.Value = value.toLowerCase().trim() }

  Valid(): Error[] {
    let errs: Error[] = []
    if (this.Value.length < 3) { errs.push(toShortNameError) }
    if (this.Value.length > 64) { errs.push(toLongNameError) }
    if (!nameRegex.test(this.Value)) { errs.push(invalidCharachtersNameError) }
    return errs
  }

  Errors(): Error[] {
    return [toShortNameError, toLongNameError, invalidCharachtersNameError]
  }
}

// password

const smallLeterRegex = new RegExp("[a-z]")
// const bigLeterRegex = new RegExp("[A-Z]")
const numberRegex = new RegExp("[0-9]")
// const specialCharacterRegex = new RegExp("[!@#$%^&*()_+{}:\"|<>?]")


const toShortPasswordError = new Error("password has to be at least 3 characters long")
const toLongPasswordError = new Error("password cannot have more than 64 characters")
const noSmallCharactersPasswordError = new Error("password must contain small letter")
// const noBigCharactersPasswordError = new Error("password must contain big letter")
const noNumbersPasswordError = new Error("password must contain number")
// const noSpecialCharactersPasswordError = new Error("password must contain special character")

// export const PasswordInputProps: { [key: string]: any } = {
//   type: "password",
//   placeholder: "********",
// }
export class Password implements VO {
  public Value: string

  constructor(value: string) {
    this.Value = value
  }

  Valid(): Error[] {
    let errs: Error[] = []
    if (this.Value.length < 3) { errs.push(toShortPasswordError) }
    if (this.Value.length > 64) { errs.push(toLongPasswordError) }
    if (!smallLeterRegex.test(this.Value)) { errs.push(noSmallCharactersPasswordError) }
    // if (!bigLeterRegex.test(this.Value)) { errs.push(noBigCharactersPasswordError) }
    if (!numberRegex.test(this.Value)) { errs.push(noNumbersPasswordError) }
    // if (!specialCharacterRegex.test(this.Value)) { errs.push(noSpecialCharactersPasswordError) }
    return errs
  }

  Errors(): Error[] {
    return [
      toShortPasswordError,
      toLongPasswordError,
      noSmallCharactersPasswordError,
      // noBigCharactersPasswordError,
      noNumbersPasswordError,
      // noSpecialCharactersPasswordError,
    ]
  }
}

// login

export class Login implements VO {
  public Value: string

  constructor(value: string) {
    this.Value = new Name(value).Value
  }

  Valid(): Error[] {
    return new Name(this.Value).Valid()
  }

  Errors(): Error[] {
    return new Name(this.Value).Errors()
  }
}

File: api/application/modules/usersmodule/valueobjects.go
package usersmodule

import (
	"database/sql/driver"
	"encoding/json"
	"errors"
	"regexp"
	"strings"
)

// user name
// user password

// user name

type UserName string

func (u *UserName) UnmarshalJSON(data []byte) error {
	var s string
	err := json.Unmarshal(data, &s)

	if err != nil {
		return err
	}

	*u = UserName(strings.Trim(strings.ToLower(s), " "))

	return nil
}

func (UserName) GormDataType() string { return "varchar(64)" }

var nameRegex = regexp.MustCompile("^[a-zA-Z0-9-_]*$")

func (name *UserName) Valid() []error {
	var errs []error
	if len(*name) < 3 {
		errs = append(errs, errors.New("name has to be at least 3 characters long"))
	}
	if len(*name) > 64 {
		errs = append(errs, errors.New("name cannot have more than 64 characters"))
	}
	if ok := nameRegex.MatchString(string(*name)); !ok {
		errs = append(errs, errors.New("name can contain only letters, numbers, - and _"))
	}
	return errs
}

// user password

type UserPassword string

var smallLeterRegex = regexp.MustCompile("[a-z]")

// var bigLeterRegex = regexp.MustCompile("[A-Z]")
var numberRegex = regexp.MustCompile("[0-9]")

// var specialCharacterRegex = regexp.MustCompile("[!@#$%^&*()_+{}:\"|<>?]")

func (password *UserPassword) Valid() []error {
	var errs []error
	if len(*password) < 3 {
		errs = append(errs, errors.New("password has to be at least 3 characters long"))
	}
	if len(*password) > 64 {
		errs = append(errs, errors.New("password cannot have more than 64 characters"))
	}
	if ok := smallLeterRegex.MatchString(string(*password)); !ok {
		errs = append(errs, errors.New("password must contain small letter"))
	}
	// if ok := bigLeterRegex.MatchString(string(*password)); !ok {
	// 	errs = append(errs, errors.New("password must contain big letter"))
	// }
	if ok := numberRegex.MatchString(string(*password)); !ok {
		errs = append(errs, errors.New("password must contain number"))
	}
	// if ok := specialCharacterRegex.MatchString(string(*password)); !ok {
	// 	errs = append(errs, errors.New("password must contain special character"))
	// }
	return errs
}

func (UserPassword) GetValue() (driver.Value, error) {
	return nil, errors.New("cannot store raw password in database")
	// return string(password), nil
}
